### ISSUE TITLE
Implement Workflow for building releases (both beta and official)

### ISSUE THREAD

**Issue by derimmer** [vasl-developers/vasl#1878]
@geezer09 I am ready to go with trying your workflow to build the 6.7.0-beta1 release and all the related steps. 

Let me know how you want to proceed.

**Comment by geezer09** [comment]
Hey @derimmer, my workflows are pretty much ready to test for beta releases. I guess the only difference with beta is they don’t get published to  the website right? I still need a bit of work to automate the website update but probably wont need that for a little while

Do we need to upload the compressed repo files to the release page? Currently I am only uploading the .vmod file. Why would anyone want the source code, and if they do they can just pull it off the repo at the tag associated with the release

**Comment by derimmer** [comment]
> Hey @derimmer, my workflows are pretty much ready to test for beta releases. I guess the only difference with beta is they don’t get published to the website right? I still need a bit of work to automate the website update but probably wont need that for a little while
> 
Correct on both counts

> Do we need to upload the compressed repo files to the release page? Currently I am only uploading the .vmod file. Why would anyone want the source code, and if they do they can just pull it off the repo at the tag associated with the release
No, I have never uploaded the compressed repo files. Github seems to do that all by itself. 

Feel free to go ahead and trigger the release. 

Then show me how you do it. Thanks.

**Comment by geezer09** [comment]
Okay I ran the workflow and it seem to work for the most parts, still needs some tweaking but getting there. I left the release as a draft, if you want to have a quick look at the release notes and and then you can publish if you are happy with it.

Here is a brief description of how the workflow operates. There are alot of little check and steps that will probably need to be fine tuned and changed but right now here is what is going on under the hood. As you might have seen I have taken the master branch and renamed it to production, the idea is that every single commit to production will produce a new release, this will give us a clear time line of what happened from release to release and will help us when debugging. This will not be 100% true at the start because I will have to make plenty of little of adjustments to the workflow files, which will produce commits without new versions, but we'll get there.

In order to push a commit to production, the proper way of doing that if from the develop branch. Pull any work done since last release into develop, then create a temporary branch called (for example) release 6.7.5. This branch is not used to do any "useful work" your just preparing it to merge into production, this means you are updating the new release-notes.md file i have place in the dist folder to include all relevant documentation for this new release (bug fixes, new features, etc). You also need to upload the new version of the anticipated release in the pom.xml file.

Once this is done and you are ready to merge, you create a pull request from your temporary branch (release 6.7.5) to production. This will start a series of "pre-test" to make sure you are indeed ready to release a new version. If you look in the .github/workflow folder you will see the three workflow files responsible for all of the automation. PR_check runs when a pull request is triggered and you can take a a look to see what it checks for but basically it makes sure you incremented the version properly, does a test build to make sure the maven package command dosent produce any errors and runs the new script i wrote to check to make sure that map versions match. 

In this pull request, if any of the test fails, look at the output and fix them until all test pass. At that point, merge your pull request and then the create_release workflow triggers. If everything works well, this will create a new release with the proper version number, upload the .vmod file to the release and display the release-notes.md to the release page. 

The last step is to update the website, im currenlty running into a few problems there but shold have it all sorted out well before we need to update the website. If the version you selected has "beta" in it, then this step is automatically skipped anyways.

After the release is produced, a post to BGG and wherever else you desire is all that should be required. I also recommend pulling in any changes from production into develop and then starting all over by branching from develop for the next iteration. Ideally when you start a new iteration you should only have two or three branches. Production, develop and devx.y.z or something like that.   

At this point I have kept the publication of the release in draft form so that modifications can easily be made to it and manual approval of the website update will be required until we iron out the kinks and can let 'er rip. Once we do make things run smoother. ill create a document in the repo to outline these steps and any additional ones we deem necessary.

Hope this helps and don't hesitate if you have any questions or comments.

**Comment by derimmer** [comment]
> . . .  does a test build to make sure the maven package command dosent produce any errors . . .
> 
So, we continue, at some point to use the "mvnw" command to compile the .vmod?
 
> After the release is produced, a post to BGG and wherever else you desire is all that should be required. I also recommend pulling in any changes from production into develop and then starting all over by branching from develop for the next iteration. Ideally when you start a new iteration you should only have two or three branches. Production, develop and devx.y.z or something like that.
>
Can we automate branch descruction as part of the above? I think one of our issues is stale branches hanging around forever. I agree with the goal of starting from as clean a sheet as possible after each release: (1) reset develop from production, (2) delete all branches that do not have commits that are not in develop - or maybe there is a better test?
> 
> At this point I have kept the publication of the release in draft form so that modifications can easily be made to it and manual approval of the website update will be required until we iron out the kinks and can let 'er rip. Once we do make things run smoother. ill create a document in the repo to outline these steps and any additional ones we deem necessary.
> 
> Hope this helps and don't hesitate if you have any questions or comments.
>
This is excellent as far as I am concerned.

**Comment by derimmer** [comment]
I see you have published the release. Really like the presentation! Big improvement. 

I will post on FB and Discord. 

I will update the VASL Coop page on VASL.info

**Comment by derimmer** [comment]
I downloaded and started the beta1.vmod. All worked as expected. 

I did note that the What's New file was not updated inside the beta1. That's a step we will need to integrate in the next iteration. In my view, the workflow should check that the What's New file has been updated since the previous release. If not, it should trigger an error. Would it be possible to have the release-notes.md turned into a pdf by the workflow? I have no concern about preserving the current format of the What's New document - I kind of like your format better.

**Comment by geezer09** [comment]
Yeah i’ve been working on that issue. The pdf file format is a bit unfortunate because it does  not lend itself well to the ci/cd pipeline. I’ll come up with a solution shortly, i’d like the file accessible from within the game to be a bit more comprehensive then what is shown on the release page, perhaps show a history of the last few versions. 

Still a work in progress.

**Comment by derimmer** [comment]
I noticed that create_release.yaml invokes mvn clean package on line 39. Recently, we have been using 'mvnw clean package' to build the module. 

I have been trying to update branch feature/vasldougreg by pulling all the 'develop' changes into it. It is capturing the following problems:

![image](https://github.com/user-attachments/assets/9d6ec8fb-4f0f-48f1-9fc8-5c04d59bafbb)

There are about 20+ more files in addition to the ones above. I do not know how to resolve these errors. Advice?

**Comment by geezer09** [comment]
Noted about the mvnw command, ill update that for next release.

As for the above ‘errors’ they are just merge conflicts. This happens when you try to merge two branches where changes were made to same lines within a file. You just have to manually go through those and decide what you want to keep and what you want to throw. That tends to happen with older branches. The process is a bit confusing at first but nothing destructive is done so don’t worry too much about.

**Comment by derimmer** [comment]
I am having trouble with the "manually go through those". How do I actually do that? In all of these cases, a file needs to be deleted from the branch I am trying to pull changes INTO.

Do I close the PR, delete the files and rerun the PR? I have tried that and I can't seem to get the files to delete from within Github. 

Should I be doing it another way?

Sorry to bug you with what should be a simple process.

**Comment by geezer09** [comment]
I wouldnt do this as a PR. I’m not sure exactly what your trying to do, but merge development into your feature branch locally and then intellij will help you sort out the conflicts. Once it is done in your local repo you can push that up to the remote.

**Comment by derimmer** [comment]
Duh! I have only done that about a 100 times! Why do it the easy way when there is a more complex option!

Done and dusted, I will delete the github branch.

**Comment by derimmer** [comment]
Just used this process to release vasl-6.7.0-beta4.vmod. All went well. Having done this a couple of times now, it begins to feel pretty straightforward. I remembered to update the release-notes.md.

However, 6.70-beta4 is not showing up on the releases page. Do I remember correctly that I have to do something on that page to actually publish it? I can't even find a reference to a draft. I think I am missing something somewhere.

**Comment by derimmer** [comment]
I have found the problem I think.

**Comment by derimmer** [comment]
The build workflow is working and I get the "tests passed" messages. I had not updated the release-notes.md from beta3 to beta4 and that was breaking the create-release workflow. Once I did that, it worked and beta4 is now on the github releases page.

Note: the create-release workflow does not seem to report out anywhere on github that I can find. I eventually get an email but there is a lag. Where should I be looking on github?

**Comment by geezer09** [comment]
Check the actions tabs in github for details about the workflows and their outcomes

**Comment by derimmer** [comment]
I see them now. thanks. Didn't know where to look.

I think the work you have done implementing workflows is super; a big step forward.

**Comment by geezer09** [comment]
Hey @derimmer, I saw you just released a new beta. A few comments.

1) Something is not quite right. Every single commit within the Production branch should be it’s own fully functional and working release version, seems like this is not the case. Can you explain what happened? One thing I mentioned a while back but should now bring back up is that the production branch should be protected such that only commits can only be made to it via pull requests. I’ll let you do that since I don’t have admin rights.

2) The commit history is a bit messy and hard to follow. For example I see a series of ten or so commits with the exact same name. I do that all the time when working on my local machine, but it is good practice to try and do a bit of a cleanup before pushing to the remote repository. Git has many handy commands that allow you to clean up your commit history before pushing to the remote.  It only takes a minute and is good practice. Be warned though, do not attempt to clean the history of commits once they have been pushed to the remote, that is always a nightmare.

3) I think what might be happening is when you are trying to create a new release and you don’t pass all the test, you create a new PR. You don’t have to do that, simply create a new commit on your current PR that didnt pass all the test and that will re-trigger the workflows and testing. 

Hope that makes sense, let me know if you have any questions or suggestions for the workflow. Also what happened to beta5? Seems like we skipped one. If you give me a heads ip before the next beta ill make sure my changes get incorporated as well.

**Comment by derimmer** [comment]
> Hey [@derimmer](https://github.com/derimmer), I saw you just released a new beta. A few comments.
> 
> 1. Something is not quite right. Every single commit within the Production branch should be it’s own fully functional and working release version, seems like this is not the case. Can you explain what happened? One thing I mentioned a while back but should now bring back up is that the production branch should be protected such that only commits can only be made to it via pull requests. I’ll let you do that since I don’t have admin rights.

I get more confident and better able to use it each time I try the workflow. It is an excellent tool. Thanks again for building it. In this instance, each of my commits to Production was indeed a full release version. The first time I got an error. I was able to drill in and find more info on the nature of the error than I had before. So, I made a change locally to fix, pushed it to develop and went through the process again. I might of had a second error and had to do it a third time, I can't remember. In the end, everything worked, and I was able to see the beta6 entry in releases, which I changed from a draft to a published pre-release. As far as I was concerned, this worked as expected and as per your previous instructions. 

I will add the protection to Production. 

> 2. The commit history is a bit messy and hard to follow. For example I see a series of ten or so commits with the exact same name. I do that all the time when working on my local machine, but it is good practice to try and do a bit of a cleanup before pushing to the remote repository. Git has many handy commands that allow you to clean up your commit history before pushing to the remote.  It only takes a minute and is good practice. Be warned though, do not attempt to clean the history of commits once they have been pushed to the remote, that is always a nightmare.

You are referring to my commits from the local environment to my dev branch in github. I agree that they are messy. That is mainly because my coding process is messy! In this particular case, the problem was magnified because I was trying to manage the addition of about 6000 new image files, which my IDE would not let me do as one fell swoop. I did what I could to get this done. By the time I finished, I had come up with a better strategy. Still, I take your point and will explore the git commands as you suggest.

> 3. I think what might be happening is when you are trying to create a new release and you don’t pass all the test, you create a new PR. You don’t have to do that, simply create a new commit on your current PR that didn't pass all the test and that will re-trigger the workflows and testing. 

As per the above, you are totally correct. I am not at all confident about "creating a new commit on my current PR" but I will give it a try next time.

> Hope that makes sense, let me know if you have any questions or suggestions for the workflow. Also what happened to beta5? Seems like we skipped one. If you give me a heads ip before the next beta ill make sure my changes get incorporated as well.

There is a beta5 which has a particular set of changes. I sent it to the requestor to see if it met requirements and haven't heard back so it and the changes it contains have not yet been publicly released.  The major issue is that every time we change a counter in VASL we have to upload the buildFile to github. And as you know, the buildFile changes are not merged into the existing file. The new version of the buildFile replaces the existing buildFile, potentially overwriting changes from other commits (from multiple users). That was very much the risk with the integration of the svg images, which we chose to do a block at a time for good and legitimate reasons. 

I will try to remember to put out a call when I am about to build a version. That makes sense.
